xray/backends/pynio.py

```diff
@@ -0,0 +1,247 @@
+                                  "this is not supported by the netCDF4 python "
+                                  "library.")
+    return var
+
+
+class NioDataStore(AbstractWritableDataStore):
+    """Store for reading and writing data via the PyNIO library.
+
+    This store supports NetCDF3, NetCDF4 and OpenDAP datasets.
+    """
+    def __init__(self, filename, mode='r', clobber=True, diskless=False,
+                 persist=False, format='NETCDF4', group=None):
+        import Nio
+        ds = Nio.open_file(filename, mode=mode)
+
+#, clobber=clobber,
+#                         diskless=diskless, persist=persist,
+#                         format=format)
+        with close_on_error(ds):
+            self.ds = _nc4_group(ds, group, mode)
+        self.format = format
+        self._filename = filename
+
+    def store(self, variables, attributes):
+        # All NetCDF files get CF encoded by default, without this attempting
+        # to write times, for example, would fail.
+        cf_variables, cf_attrs = cf_encoder(variables, attributes)
+        AbstractWritableDataStore.store(self, cf_variables, cf_attrs)
+
+    def open_store_variable(self, var):
+        #var.set_auto_maskandscale(False)
+        dimensions = var.dimensions
+        data = indexing.LazilyIndexedArray(NetCDF4ArrayWrapper(var))
+        attributes = OrderedDict((k, var.attributes[k])
+                                 for k in var.attributes.keys())
+        _ensure_fill_value_valid(data, attributes)
+        # netCDF4 specific encoding; save _FillValue for later
+        encoding = {}
+        #filters = var.filters()
+        filters = None
+        if filters is not None:
+            encoding.update(filters)
+        if hasattr(var,'chunk_dimensions'):
+            chunking = var.chunk_dimensions
+        else:
+            chunking = None
+        if chunking is not None:
+            if chunking == 'contiguous':
+                encoding['contiguous'] = True
+                encoding['chunksizes'] = None
+            else:
+                encoding['contiguous'] = False
+                encoding['chunksizes'] = tuple(chunking)
+        # TODO: figure out how to round-trip "endian-ness" without raising
+        # warnings from netCDF4
+        # encoding['endian'] = var.endian()
+        #pop_to(attributes, encoding, 'least_significant_digit')
+        # save source so __repr__ can detect if it's local or not
+        encoding['source'] = self._filename
+        return Variable(dimensions, data, attributes, encoding)
+
+    def get_variables(self):
+        return FrozenOrderedDict((k, self.open_store_variable(v))
+                                 for k, v in iteritems(self.ds.variables))
+
+    def get_attrs(self):
+        print self.ds.attributes
+        return FrozenOrderedDict((k, self.ds.attributes[k])
+                                 for k in self.ds.attributes.keys())
+
+    def get_dimensions(self):
+        return FrozenOrderedDict((k, len(v))
```

Comment:
is `len(v)` correct? My own personal, incomplete implementation (created a year ago) has it be just `v`.